### PR TITLE
Fast reboot is not restoring the ARP entries learnt on Port Routing interfaces.

### PR DESCRIPTION
**Pull request by kirankella**
**- What I did**
Following bugs are fixed in fast-reboot-dump.py script
  - The ARP entries learned only on VLAN routing interfaces are being saved to /host/fast-reboot/arp.json. Fix is done to save ARP entries learned on Port routing interfaces too.
    
  - Before fast rebooting, ARP requests are being sent out for all the ARP entries, probably to refresh the ARP cache of the neighboring hosts/routers. But the ARP requests on the VLAN routing interfaces are not actually going out with the VLAN tag as intended. As a result the ARP requests don't really reach the neighbors. Fix done in the arp transmit code.

**- How I did it**
Changes done in fast-reboot-dump.py to iterate over NEIGH_TABLE entries learned on Port routing interfaces too while saving them to arp.json.
Send ARP requests with the VLAN tag for the ARP entries learned on that VLAN.

**- How to verify it**
Before and after fast-reboot, dump the ARP table in APP_DB redis (redis-cli -n 0 --scan --pattern NEIGH_TABLE*) and in the hardware (bcmcmd 'l3 l3table show').

**- Previous command output (if the output of a command-line utility has changed)**

**- New command output (if the output of a command-line utility has changed)**

-->

### CODE REVIEW

File: scripts/fast-reboot-dump.py [hunk 1]
```diff
@@ -23,17 +23,22 @@ def generate_arp_entries(filename, all_available_macs):
     keys = db.keys(db.APPL_DB, 'NEIGH_TABLE:*')
     keys = [] if keys is None else keys
     for key in keys:
-        vlan_name = key.split(':')[1]
+        key1 = key.split(':')[1]
```
Comment (pavel-shirshov) [inline scripts/fast-reboot-dump.py]: Please use meaningful name here. Like 'interface_name' for example

File: scripts/fast-reboot-dump.py [hunk 2]
```diff
@@ -199,14 +204,37 @@ def send_arp(s, src_mac, src_ip, dst_mac_s, dst_ip_s):
 
     return
 
+def send_vlan_arp(s, src_mac, src_ip, dst_mac_s, dst_ip_s, vlan_iface):
+    # convert dst_mac in binary
+    dst_ip = socket.inet_aton(dst_ip_s)
+    vlan   = int(vlan_iface[len('Vlan'):])
+    vlan_hex = '%04x' % vlan
+
+    # convert dst_ip in binary
```
Comment (pavel-shirshov) [inline scripts/fast-reboot-dump.py]: mac, not ip

File: scripts/fast-reboot-dump.py [hunk 3]
```diff
@@ -23,17 +23,22 @@ def generate_arp_entries(filename, all_available_macs):
...
         ip_addr = key.split(':')[2]
         entry = db.get_all(db.APPL_DB, key)
-        if (vlan_name, entry['neigh'].lower()) not in all_available_macs:
-            # FIXME: print me to log
+        if key1.startswith('eth0') or key1.startswith('lo'):
             continue
+        if entry['family'] == 'IPv6':
+            continue
+        if key1.startswith('Vlan'):
+            if (key1, entry['neigh'].lower()) not in all_available_macs:
```
Comment (pavel-shirshov) [inline scripts/fast-reboot-dump.py]: may be change to ifs to one?

File: scripts/fast-reboot-dump.py [hunk 4]
```diff
@@ -199,14 +204,37 @@ def send_arp(s, src_mac, src_ip, dst_mac_s, dst_ip_s):
 
     return
 
+def send_vlan_arp(s, src_mac, src_ip, dst_mac_s, dst_ip_s, vlan_iface):
```
Comment (pavel-shirshov) [inline scripts/fast-reboot-dump.py]: I think it's better to make one function to send an arp, but with argument dot1Qtag=False by default.

File: scripts/fast-reboot-dump.py [hunk 5]
```diff
@@ -216,9 +244,13 @@ def garp_send(arp_entries, map_mac_ip_per_vlan):
         sockets[src_if].bind((src_if, 0))
 
     # send arp packets
-    for vlan_name, dst_mac, dst_ip in arp_entries:
-        src_if = map_mac_ip_per_vlan[vlan_name][dst_mac]
-        send_arp(sockets[src_if], src_mac_addrs[src_if], src_ip_addrs[vlan_name], dst_mac, dst_ip)
+    for iface_name, dst_mac, dst_ip in arp_entries:
+        if iface_name in map_mac_ip_per_vlan:
+            src_if = map_mac_ip_per_vlan[iface_name][dst_mac]
+            send_vlan_arp(sockets[src_if], src_mac_addrs[src_if], src_ip_addrs[iface_name], dst_mac, dst_ip, iface_name)
```
Comment (pavel-shirshov) [inline scripts/fast-reboot-dump.py]: Anyway I don't understand why we need to send arp packets with dot1Q tags on non-trunk vlan interfaces?

**Comment by pavel-shirshov** [comment]
It's not intended to send packets with VLAN tag on untagged vlan interfaces.

**Comment by pavel-shirshov** [comment]
Please fix other comments.

File: scripts/fast-reboot-dump.py (hunk 3 above)
Comment (kirankella) [reply scripts/fast-reboot-dump.py]: Done

File: scripts/fast-reboot-dump.py (hunk 4 above)
Comment (kirankella) [reply scripts/fast-reboot-dump.py]: Done

**Comment by kirankella** [comment]
> It's not intended to send packets with VLAN tag on untagged vlan interfaces.
We do not know the tagging/untagging configuration of an interface in the VLAN in the current code.
Did you mean to say, query the tagging mode for the interface in a VLAN from the configdb and then send ARP with or without tag?

**Comment by kirankella** [comment]
> > It's not intended to send packets with VLAN tag on untagged vlan interfaces.
> 
> We do not know the tagging/untagging configuration of an interface in the VLAN in the current code.
> Did you mean to say, query the tagging mode for the interface in a VLAN from the configdb and then send ARP with or without tag?

**Comment by kirankella** [comment]
> Please fix other comments.

@pavel-shirshov Can you please review the latest changes in this PR. All comments are addressed.

**Pull request closed without merging**